tests/kernel/common/src/printk.c

```diff
@@ -16,6 +16,16 @@ void *__printk_get_hook(void);
 int (*_old_char_out)(int);
 
 #if defined(CONFIG_PICOLIBC)
+
+#ifdef _WANT_IO_LONG_LONG
+/* long long support present if _WANT_IO_LONG_LONG is defined */
+#define HAS_LONG_LONG
+#endif
+#if !defined(PICOLIBC_INTEGER_PRINTF_SCANF) && !defined(FORMAT_DEFAULT_INTEGER)
+/* long long support present if floating point output is in use */
+#define HAS_LONG_LONG
+#endif
```

Comment:
> This ensures that we detect the actual picolibc configuration rather than
relying on the local requested configuration values as those may
differ when using the toolchain version of picolibc.

I think what we want to do is to (de-)select `PICOLIBC_IO_LONG_LONG` and `PICOLIBC_IO_FLOAT` when the toolchain picolibc is used so that `CONFIG_PICOLIBC_IO_LONG_LONG` and `CONFIG_PICOLIBC_IO_FLOAT` are always correct and be relied upon.

We can also add sanity checks for verifying that these Kconfigs match the actual picolibc configs in the `libc-hooks.c`.

Reply:
These two config options are slightly different here, and you're at least half right about this case :-)

`CONFIG_PICOLIBC_IO_LONG_LONG` is a build-time option in picolibc which removes `long long` support from printf and scanf. That can only work when building picolibc as a module; it's not something Zephyr code should be using as it cannot be reliably set when using the toolchain version unless we were to add Kconfig-time support for looking into the picolibc.h header file.

On the other hand, `CONFIG_PICOLIBC_IO_FLOAT` affects the module-version of picolibc differently from the toolchain version, in both cases making printf (et al) support floating point operands. So,yeah, need to use the CONFIG define for this piece.

Reply:
Hrm. Should probably just make `CONFIG_PICOLIBC_IO_LONG_LONG` only settable when building picolibc as a module then?

Reply:
Let me know if you accept this solution, or if we need to do something fancier with how long long support is handled.

Reply:
> `CONFIG_PICOLIBC_IO_LONG_LONG` is a build-time option in picolibc which removes `long long` support from printf and scanf. That can only work when building picolibc as a module; it's not something Zephyr code should be using as it cannot be reliably set when using the toolchain version unless we were to add Kconfig-time support for looking into the picolibc.h header file.

I was thinking more along the lines of hard-coding its value when the toolchain picolibc is used with the assumption that the toolchain being used is Zephyr SDK, and eventually moving on to resolving these configs in CMake once we have the infrastructure to do that; but, using the `_WANT_IO_LONG_LONG` directly here does ensure that it behaves correctly all the time.

The only concern is that the mismatch between `CONFIG_PICOLIBC_IO_LONG_LONG` and `_WANT_IO_LONG_LONG` can lead to potential issues in the future (e.g. if someone who is not an expert on this matter comes and adds a check using `CONFIG_PICOLIBC_IO_LONG_LONG` assuming it is always representative of the long long IO support in picolibc), but we can fix that issue incrementally later.